reference/sql_workarounds.rst

```diff
@@ -0,0 +1,145 @@
+
+Subqueries
+----------
+
+In analytics it is common to model information using a "star schema." This is a collection of dimensional tables (e.g. observations, statistics) referencing a table of facts (e.g. people, pages or events). Furthermore, a common form of query examines dimensions for a subset of the facts. Examples of these kind of queries:
+
+* "Get aggregated statistics about football players whose passes exceed three hundred yards."
+* "Which other pages do people visit who view a certain page?"
+* "Count the number of distinct sessions for the top twenty-five most visited blog posts."
+* "What is the number of people in companies among the top fifty product areas?"
+
+The general form of the query in SQL is
+
+.. code-block:: sql
+
+  select dim_a, dim_b, dim_c
+    from dimensions
+   where fact_id in (select id from facts where ...conditions...)
+     and ...conditions...;
+
+Citus does not allow subqueries in the WHERE clause so we must choose a workaround.
+
+Workaround 1. Generate explicit WHERE-IN expression
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Pros
+    * Able to distribute the query across shards
+* Cons
+    * Constructed query not suitable when facts subquery returns large number of results
+    * Somewhat hard to read, requires building custom SQL
+
+In this technique we use PL/pgSQL to construct and execute one statement based on the results of another.
+
+.. code-block:: sql
+
+  -- create temporary table with results
+  do language plpgsql $$
+    declare facts_array integer[];
+  begin 
+    execute 'select id from facts where ...conditions...'
+       into facts_array;
+    execute format(
+      'create temp table results_temp as '
+      'select dim_a, dim_b, dim_c'
+      '  from dimension'
+      ' where fact_id = any(array[%s])'
+      '   and ...conditions...',
+      array_to_string(facts_array, ','));
+  end;
+  $$;
+
+  -- read results, remove temp table
+  select * from results_temp;
+  drop table results_temp;
+
+
+Workaround 2. Duplicate on Master
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Pros
+    * Works for subqueries returning any number of results
+* Cons
+    * Must transmit full rows from the queries back to the master
+    * Requires extra storage space in the master database
+
+In this workaround the client runs the outer- and sub-query independently, saves their results, and joins them.
+
+.. code-block:: sql
+
+  -- Capture the dimension query results
+  create temp table dim_temp as
+  select fact_id, dim_a, dim_b, dim_c
+    from dimensions
+   where ...conditions...;
+  
+  -- Capture the subquery results
+  create temp table fact_temp as
+  select id
+    from facts
+   where ...conditions...;
+  
+  -- Run the query on local tables where subqueries are OK
+  select dim_a, dim_b, dim_c
+    from dim_temp
+   where fact_id in (select id from fact_temp);
+
+  -- Remove temp tables
+  drop table dim_temp;
+  drop table fact_temp;
+
+SELECT ... INTO
```

Comment:
I guess the standard query is INSERT INTO SELECT. Refer the [insert documentation](https://www.postgresql.org/docs/9.5/static/sql-insert.html) of postgresql 9.5 which talks about VALUES in insert could be taken from a SQL query.